Reworking iree_status_t/iree::Status and improving error messages across the VM

benvanik opened the pull request:
This fixes #265 by implementing the described iree_status_t behavior and changing Status/StatusBuilder/StatusOr to wrap it. By doing this I was able to remove quite a bit of gunk that was required to move between the Google-style C++ Status and the API (such as ToApiStatus/FromApiStatus).

By being able to integrate iree_status_t into the C++ types I was also able to unify all of the macros - so IREE_CHECK_OK and IREE_RETURN_IF_ERROR can take either iree_status_t or Status/etc. Because the behavior is now different we can no longer use the iree/base/google/ versions and as such I've renamed all macros such that there are no collisions with the non-IREE versions.
Now, IREE_CHECK_OK/IREE_RETURN_IF_ERROR when used in C code can take optional printf-style arguments to add annotations to results, for example: `IREE_RETURN_IF_ERROR(OtherFunc(...), "with a value: %d", 5);`
When used in C++ the macros still accept printf-style arguments but can also use the ostream style: `IREE_RETURN_IF_ERROR(OtherFunc(...)) << "with a value: " << 5;`. Though it's supported one shouldn't mix them!

The most interesting changes in this branch are in iree/base/api.h + iree/base/api.c and the iree/base/internal/status* files. Other changes are mechanical cleanups required to avoid name conflicts, enable C compilation of core parts (or code to be used both in C and C++), and removing abseil deps from files that are C compatible.

A lot of focus was put on keeping the status usage - even if source location, messages, and stack traces are attached - from increasing codesize or overhead when on the happy success path. https://gcc.godbolt.org/z/xGsPc3

For example, this chain of calls and checks:
![image](https://user-images.githubusercontent.com/75337/89965402-91f43480-dc01-11ea-8514-d307828f4910.png)

Ends up as the minimal possible code (while still having any kind of success checks):
![image](https://user-images.githubusercontent.com/75337/89965439-aafce580-dc01-11ea-9359-fade63cac4ba.png)

The error handling cases are moved out of the way and (hopefully) kept out of cache. The handlers are able to hit the tail call optimization happy path and keeps the whole function code size smaller:
![image](https://user-images.githubusercontent.com/75337/89965512-e697af80-dc01-11ea-9b2c-1c42aabbb788.png)

benvanik committed:
Improving IREE C status macros and support functions.
Progress on #265.

benvanik committed:
Changing IREE_ALLOCATOR_* macros to static inline functions.

benvanik committed:
clang-tidy fixes for VM header files (mostly macros->static inline).

benvanik committed:
Removing IREE_API_NO_PROTOTYPES as it wasn't consistently used
(only needed when someone wants to use us as a dylib with dlopen)

benvanik committed:
Making more of iree/base/api.cc -> c

benvanik committed:
Fixing type usage in iree_string_view_* functions.

benvanik committed:
Removing file mapping from the C API.

benvanik committed:
Converting iree_string_view_match_pattern -> c

benvanik committed:
Adding iree_string_view_remove_prefix to round out required set for HAL.

benvanik committed:
Adding IREE_STATUS_FEATURE/IREE_STATUS_MODE.

benvanik committed:
Removing unused InitializeEnvironment from API.
//iree/base:main takes care of it and keeps us from always pulling in
flags/absl/etc.

benvanik committed:
iree/base/api.cc -> api.c (besides iree_time_now()).

benvanik committed:
Removing iree/base/source_location.h.
Future changes will likely remove it entirely (or simplify it).

benvanik committed:
Removing/moving some rarely used utilities.

benvanik committed:
Merging iree/base/platform_headers.h -> target_platform.h

benvanik committed:
Updating iree_status_t usage and adding tons of error messages.
Progress on #265.

benvanik committed:
Adding status messages for all C status results.

benvanik committed:
bytecode_module.cc -> bytecode_module.c

benvanik committed:
Changing size values to iree_host_size_t (size_t).

benvanik committed:
Replacing some VM macros with static inline functions.

# Conflicts:
#	iree/vm/value.h

benvanik committed:
Forcing status payload type to unsigned.

benvanik committed:
Removing usage of *Error() status builders.

benvanik committed:
Moving status_errors.h into status.h.

benvanik committed:
Removing usage of iree::Annotate.

benvanik committed:
Removing absl flags usage from Status.

benvanik committed:
Simplifying source_location.

benvanik committed:
Moving status_matchers to iree/testing/.

benvanik committed:
Rebasing StatusCode to iree_status_code_t.

benvanik committed:
Preparing for removal of Status::message and simplifying comparisons.

benvanik committed:
Removing more unused status features.

benvanik committed:
Fixing warnings found building with clang.

benvanik committed:
Simplifying status_errno and status_win32_errors.

benvanik committed:
Preventing some Status copies.

benvanik committed:
Simplifying status_macros.h

benvanik committed:
Removing FromApiStatus and unifying IREE_RETURN_IF_ERROR macros.

benvanik committed:
Renaming RETURN_IF_ERROR to IREE_RETURN_IF_ERROR.
Including the C++ file will enable support in the macro for Status.

benvanik committed:
Renaming ASSIGN_OR_RETURN to IREE_ASSIGN_OR_RETURN.

benvanik committed:
Switching out some absl macros and utils for Status use.

benvanik committed:
Status tweaks to reduce binary size.

benvanik committed:
Adding option to start fleshing out size-optimized flags for #898.

benvanik committed:
Adding IREE_ARRAYSIZE to replace ABSL_ARRAYSIZE.

benvanik committed:
Implementing iree_status_t storage and payloads.

benvanik committed:
Moving IREE_ test macros to iree/testing/status_matchers.h.

benvanik committed:
Rebasing iree::Status* on iree_status_t.

benvanik committed:
Removing ToApiStatus and api_util.h.

benvanik committed:
Replacing EXPECT/ASSERT/CHECK macros with IREE_* versions.

benvanik committed:
Removing api_time.cc now that we have a C impl.

benvanik committed:
Fixing clang/bazel warnings.

ScottTodd reviewed (commented):
First batch of comments. Still looking through.

benvanik committed:
Removing status result from ComputePoolingWindow.

ScottTodd reviewed (approved):
Please wait a bit to merge while we sync the `google`/`main` branches and coordinate the Google branch details.

benvanik committed:
Adding iree_status_from_code to avoid direct bit manipulation.

benvanik committed:
Changing iree_status_t to void* to avoid implicit conversions.
Fixes #265.

benvanik committed:
Adding fallback for a lack of C11 aligned_malloc.

ScottTodd committed:
Update base:status_test to handle IREE_STATUS_FEATURE_ANNOTATIONS. (#2917)

Otherwise the test fails in optimized builds (or other builds which disable annotations).

I used macros instead of helper functions so we get line numbers for failed test checks.

ScottTodd committed:
Fix leaks in iree/vm:stack_test. (#2922)
